File: WeiBoHomePageDemo-OC/WeiBoHomePageDemo-OC/ViewController.m
//
//  ViewController.m
//  WeiBoHomePageDemo-OC
//
//  Created by apple on 2017/7/10.
//  Copyright © 2017年 apple. All rights reserved.
//

#import "ViewController.h"
#import "MyTableView.h"
#import "MainTableViewCell.h"

@interface ViewController ()<UITableViewDataSource,UITableViewDelegate>
@property (weak, nonatomic) IBOutlet MyTableView *tableView;
@property (assign, nonatomic) BOOL  canScroll;
@property (nonatomic, strong) MainTableViewCell *userMainCell;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _canScroll = YES;//是否可以滑动
    //顶级滑动 子级不能滑动
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onOtherScrollToTop:)
                                                 name:@"kLeaveTopNtf"
                                               object:nil];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark -

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    //要减去导航栏 状态栏 以及 sectionheader的高度
    return self.view.frame.size.height-64;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return CGFLOAT_MIN;
}


- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (!self.userMainCell) {
        self.userMainCell =[tableView dequeueReusableCellWithIdentifier:@"MainTableViewCell"
                                                           forIndexPath:indexPath];
    }
    
    return self.userMainCell;
}

//子控制器到顶部了 主控制器可以滑动
- (void)onOtherScrollToTop:(NSNotification *)ntf {
    self.canScroll = YES;
    self.userMainCell.canScroll = NO;
}



- (void)scrollViewDidScroll:(UIScrollView *)scrollView{
    //子控制器和主控制器之间的滑动状态切换
    CGFloat tabOffsetY = [self.tableView rectForSection:0].origin.y-64;
    if (scrollView.contentOffset.y >= tabOffsetY) {
        scrollView.contentOffset = CGPointMake(0, tabOffsetY);
        if (_canScroll) {
            _canScroll = NO;
            self.userMainCell.canScroll = YES;
        }
    } else {
        if (!_canScroll) {
            scrollView.contentOffset = CGPointMake(0, tabOffsetY);
        }
    }
}

@end

File: WeiBoHomePageDemo-Swift/WeiBoHomePageDemo/ViewController.swift
//
//  ViewController.swift
//  WeiBoHomePageDemo
//
//  Created by apple on 2017/7/10.
//  Copyright © 2017年 apple. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    var canScroll: Bool = true
    @IBOutlet weak var mainTableView: UITableView!
    var mainCell: MainTableViewCell!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        mainTableView.dataSource = self
        mainTableView.delegate = self
        mainTableView.rowHeight = view.frame.size.height - 64
        mainTableView.tableFooterView = UIView()
        let notificationName = Notification.Name("kLeaveTopNtf")
        NotificationCenter.default.addObserver(self, selector: #selector(ViewController.onOtherScrollToTop(notificatiion:)), name: notificationName, object: nil)
        
        let gesture = UIPanGestureRecognizer(target: self, action: nil)
        view.addGestureRecognizer(gesture)
        gesture.delegate = mainTableView as? UIGestureRecognizerDelegate
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

}

extension ViewController: UITableViewDataSource, UITableViewDelegate {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if mainCell == nil {
            mainCell = tableView.dequeueReusableCell(withIdentifier: "MainCell", for: indexPath) as! MainTableViewCell
        }
        return mainCell
    }
    
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let tabOffsetY = mainTableView.rect(forSection: 0).origin.y - 64
        if scrollView.contentOffset.y >= tabOffsetY {
            scrollView.contentOffset = CGPoint(x: 0, y: tabOffsetY)
            if canScroll {
                canScroll = false
                self.mainCell.canScroll = true
            }
        }else{
            if !canScroll {
                scrollView.contentOffset = CGPoint(x: 0, y: tabOffsetY)
            }
        }
    }
    
    func onOtherScrollToTop(notificatiion: NSNotification) {
        canScroll = true
        mainCell.canScroll = false
    }
    
}
